### PR TITLE
[SPARK-14123] [SQL] DDL: Create / Drop Function

### PR DESCRIPTION
**Pull request by bomeng**
## What changes were proposed in this pull request?

This is the first step to support create / drop functions by using catalog based on @yhuai document. Instead of using NativeDDLCommand, we will call the functions in the catalog to do the jobs.
## How was this patch tested?

I've create a new test suite called DDLSuite that will test the DDL functions, eventually it will contains more DDL test cases, such as Create DB / Table, etc.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 1]
```diff
@@ -434,3 +440,57 @@ case class SetDatabaseCommand(databaseName: String) extends RunnableCommand {
...
+    functionName: String,
+    ifExists: Boolean,
+    isTemp: Boolean)(sql: String) extends RunnableCommand {
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val catalog = sqlContext.sessionState.catalog
+    val db = if (databaseName.isDefined) {
+      databaseName
+    } else {
+      Some(catalog.getCurrentDatabase)
+    }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]:
```
    val db = databaseName.getOrElse(catalog.getCurrentDatabase)
```

Then, use `Some` in the next statement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 2]
```diff
@@ -434,3 +440,57 @@ case class SetDatabaseCommand(databaseName: String) extends RunnableCommand {
...
+    functionName: String, alias: String,
+    resources: Seq[(String, String)],
+    isTemp: Boolean)(sql: String) extends RunnableCommand {
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val catalog = sqlContext.sessionState.catalog
+    val db = if (databaseName.isDefined) {
+      databaseName
+    } else {
+      Some(catalog.getCurrentDatabase)
+    }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Same rewrite here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 3]
```diff
@@ -21,15 +21,21 @@ import java.util.NoSuchElementException
...
+import org.apache.spark.sql.catalyst.CatalystTypeConverters
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.execution.SparkPlan
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.SQLContext
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]:
```scala
import org.apache.spark.sql.{Dataset, Row, SQLContext}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 4]
```diff
@@ -21,15 +21,21 @@ import java.util.NoSuchElementException
...
-import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow, TableIdentifier}
+import org.apache.spark.sql.catalyst.catalog.CatalogFunction
 import org.apache.spark.sql.catalyst.errors.TreeNodeException
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
 import org.apache.spark.sql.catalyst.plans.logical
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.CatalystTypeConverters
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.TableIdentifier
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]:
```scala
import org.apache.spark.sql.catalyst.{CatalystTypeConverters, FunctionIdentifier, InternalRow, TableIdentifier}
```

**Comment by gatorsmile** [comment]
@bomeng Overall, looks good to me. Just a few minor comments.

Can you also emphasize the following point in the `How was this patch tested?`:

- In this PR, since both `HiveContext` and `SQLContext` share the same implementation of `Create Function` and `Drop Function`, the existing test cases in `HiveContext` also verify the correctness of this PR.

Thanks for your fast work!

cc @yhuai

**Comment by bomeng** [comment]
Hive tests failed. Closed for now for more investigation.

**Pull request closed without merging**